lib/rules/generic-type-naming.js

```diff
@@ -0,0 +1,73 @@
+/**
+ * @fileoverview Enforces naming of generic type variables.
+ */
+"use strict";
+
+/**
+ * Check type parameters for rule match.
+ * @param {Array} typeParameters Nodes to be checked
+ * @param {any} context          ESLint context
+ * @param {Regex} regex          Rule name as regular expression
+ * @param {any} rule             Rule name as string
+ * @returns {void}
+ */
+function checkTypeParameters(typeParameters, context, regex, rule) {
+    const params = typeParameters && typeParameters.params;
+
+    if (!Array.isArray(params) || params.length === 0) {
+        return;
+    }
+    params.forEach(node => {
+        const type = node.type;
+
+        if (type === "TSTypeParameter" || type === "TypeParameter") {
+            const name = node.name;
+
+            if (name && !regex.test(name)) {
+                const data = { name, rule };
+
+                context.report({ node, messageId: "nameRule", data });
```

Comment:
The `messageId` API was released in [ESLint v4.15.0](https://eslint.org/blog/2018/01/eslint-v4.15.0-released#new-apis). Is it backwards-compatible? I'm wondering if we should use the older API for compatibility with all versions of ESLint:

```js
context.report({ node, message, data })
```